File: scripts/audit_knowledge_repo.py
#!/usr/bin/env python3
"""Create a deterministic health report for the Markdown knowledge corpus."""

from __future__ import annotations

import argparse
import json
import re
from collections import Counter, defaultdict
from datetime import date
from pathlib import Path

REQUIRED_FRONTMATTER = {
    "title", "date_created", "last_reviewed", "status", "source_type",
    "source_file", "doc_type", "tags",
}
FRONTMATTER = re.compile(r"\A---\s*\n(.*?)\n---\s*\n", re.DOTALL)
HEADING = re.compile(r"^(#{1,6})\s+(.+?)\s*$", re.MULTILINE)
MARKDOWN_LINK = re.compile(r"(?<!!)\[[^\]]*\]\(([^)]+)\)")


def parse_frontmatter(text: str) -> dict[str, str]:
    match = FRONTMATTER.match(text)
    if not match:
        return {}
    data = {}
    for line in match.group(1).splitlines():
        if ":" in line and not line.lstrip().startswith("#"):
            key, value = line.split(":", 1)
            data[key.strip()] = value.strip().strip('"\'')
    return data


def is_local_target(target: str) -> bool:
    return not (target.startswith(("http:", "https:", "mailto:", "#", "/")) or "://" in target)


def audit(root: Path) -> dict:
    docs = root / "docs"
    sidebar = (root / "sidebars.js").read_text(encoding="utf-8") if (root / "sidebars.js").exists() else ""
    pages = sorted(docs.rglob("*.md"))
    findings: dict[str, list[dict]] = defaultdict(list)
    titles: dict[str, list[str]] = defaultdict(list)
    doc_types = Counter()
    tags = Counter()

    for page in pages:
        relative = page.relative_to(root).as_posix()
        doc_id = page.relative_to(docs).with_suffix("").as_posix()
        text = page.read_text(encoding="utf-8", errors="replace")
        frontmatter = parse_frontmatter(text)
        if not frontmatter:
            findings["missing_frontmatter"].append({"file": relative})
        else:
            missing = sorted(REQUIRED_FRONTMATTER - frontmatter.keys())
            if missing:
                findings["incomplete_frontmatter"].append({"file": relative, "missing": missing})
            title = frontmatter.get("title", "").casefold()
            if title:
                titles[title].append(relative)
            else:
                findings["missing_title"].append({"file": relative})
            if frontmatter.get("doc_type"):
                doc_types[frontmatter["doc_type"]] += 1
            raw_tags = frontmatter.get("tags", "")
            for tag in re.findall(r"[\w-]+", raw_tags):
                tags[tag.casefold()] += 1

        headings = [(len(level), heading) for level, heading in HEADING.findall(text)]
        if not headings or headings[0][0] != 1:
            findings["invalid_h1"].append({"file": relative})
        for (previous, _), (current, heading) in zip(headings, headings[1:]):
            if current > previous + 1:
                findings["heading_level_jump"].append({"file": relative, "heading": heading, "from": previous, "to": current})
        if text.count("```") % 2:
            findings["unclosed_fence"].append({"file": relative})
        if re.search(r"\n{5,}", text):
            findings["excessive_blank_lines"].append({"file": relative})
        if "<iframe" in text.casefold():
            findings["iframe_embed"].append({"file": relative})
        if f"'{doc_id}'" not in sidebar and f'"{doc_id}"' not in sidebar:
            findings["sidebar_orphan_candidate"].append({"file": relative, "doc_id": doc_id})

        for target in MARKDOWN_LINK.findall(text):
            target = target.split("#", 1)[0].strip()
            if is_local_target(target) and target:
                candidate = (page.parent / target).resolve()
                if not candidate.exists():
                    findings["broken_relative_link"].append({"file": relative, "target": target})

    for title, paths in titles.items():
        if len(paths) > 1:
            findings["duplicate_title"].append({"title": title, "files": paths})

    return {
        "generated": date.today().isoformat(),
        "pages_scanned": len(pages),
        "document_types": dict(sorted(doc_types.items())),
        "top_tags": tags.most_common(30),
        "findings": {kind: items for kind, items in sorted(findings.items())},
        "summary": {kind: len(items) for kind, items in sorted(findings.items())},
    }


def main() -> None:
    parser = argparse.ArgumentParser()
    parser.add_argument("--root", default=".", type=Path)
    parser.add_argument("--output", type=Path)
    args = parser.parse_args()
    report = audit(args.root.resolve())
    payload = json.dumps(report, indent=2) + "\n"
    if args.output:
        args.output.write_text(payload, encoding="utf-8")
    else:
        print(payload, end="")


if __name__ == "__main__":
    main()

File: scripts/audit-knowledge-repo.mjs
#!/usr/bin/env node
/** Create a deterministic health report for the Markdown knowledge corpus. */
import fs from 'node:fs';
import path from 'node:path';

const root = process.cwd();
const outputIndex = process.argv.indexOf('--output');
const output = outputIndex >= 0 ? process.argv[outputIndex + 1] : null;
const docs = path.join(root, 'docs');
const sidebar = fs.readFileSync(path.join(root, 'sidebars.js'), 'utf8');
const required = new Set(['title', 'date_created', 'last_reviewed', 'status', 'source_type', 'source_file', 'doc_type', 'tags']);
const findings = new Map();
const titles = new Map();
const types = new Map();
const tags = new Map();

function add(kind, item) { if (!findings.has(kind)) findings.set(kind, []); findings.get(kind).push(item); }
function increment(map, key) { map.set(key, (map.get(key) ?? 0) + 1); }
function collect(directory) {
  return fs.readdirSync(directory, { withFileTypes: true }).flatMap((entry) => {
    const full = path.join(directory, entry.name);
    return entry.isDirectory() ? collect(full) : (entry.isFile() && entry.name.endsWith('.md') ? [full] : []);
  });
}
function frontmatter(text) {
  const match = text.match(/^---\s*\r?\n([\s\S]*?)\r?\n---\s*\r?\n/);
  if (!match) return {};
  return Object.fromEntries(match[1].split(/\r?\n/).flatMap((line) => {
    const i = line.indexOf(':');
    return i > 0 && !line.trimStart().startsWith('#') ? [[line.slice(0, i).trim(), line.slice(i + 1).trim().replace(/^['"]|['"]$/g, '')]] : [];
  }));
}

const pages = collect(docs);
for (const page of pages) {
  const relative = path.relative(root, page).split(path.sep).join('/');
  const docId = path.relative(docs, page).replace(/\\/g, '/').replace(/\.md$/, '');
  const text = fs.readFileSync(page, 'utf8');
  const meta = frontmatter(text);
  if (!Object.keys(meta).length) add('missing_frontmatter', { file: relative });
  else {
    const missing = [...required].filter((key) => !(key in meta)).sort();
    if (missing.length) add('incomplete_frontmatter', { file: relative, missing });
    if (meta.title) { const key = meta.title.toLowerCase(); titles.set(key, [...(titles.get(key) ?? []), relative]); }
    else add('missing_title', { file: relative });
    if (meta.doc_type) increment(types, meta.doc_type);
    for (const tag of (meta.tags?.match(/[\w-]+/g) ?? [])) increment(tags, tag.toLowerCase());
  }
  const headings = [...text.matchAll(/^(#{1,6})\s+(.+?)\s*$/gm)].map((m) => [m[1].length, m[2]]);
  if (!headings.length || headings[0][0] !== 1) add('invalid_h1', { file: relative });
  for (let i = 1; i < headings.length; i += 1) if (headings[i][0] > headings[i - 1][0] + 1) add('heading_level_jump', { file: relative, heading: headings[i][1], from: headings[i - 1][0], to: headings[i][0] });
  if ((text.match(/```/g) ?? []).length % 2) add('unclosed_fence', { file: relative });
  if (/\n{5,}/.test(text)) add('excessive_blank_lines', { file: relative });
  if (/<iframe/i.test(text)) add('iframe_embed', { file: relative });
  if (!sidebar.includes(`'${docId}'`) && !sidebar.includes(`\"${docId}\"`)) add('sidebar_orphan_candidate', { file: relative, doc_id: docId });
  for (const match of text.matchAll(/(?<!!)\[[^\]]*\]\(([^)]+)\)/g)) {
    const target = match[1].split('#')[0].trim();
    if (target && !/^(https?:|mailto:|#|\/)/.test(target) && !target.includes('://') && !fs.existsSync(path.resolve(path.dirname(page), target))) add('broken_relative_link', { file: relative, target });
  }
}
for (const [title, files] of titles) if (files.length > 1) add('duplicate_title', { title, files });
const asObject = (map) => Object.fromEntries([...map].sort(([a], [b]) => a.localeCompare(b)));
const report = {
  generated: new Date().toISOString().slice(0, 10), pages_scanned: pages.length,
  document_types: asObject(types), top_tags: [...tags].sort((a, b) => b[1] - a[1]).slice(0, 30),
  findings: asObject(findings), summary: asObject(new Map([...findings].map(([key, value]) => [key, value.length]))),
};
const payload = `${JSON.stringify(report, null, 2)}\n`;
if (output) fs.writeFileSync(path.resolve(root, output), payload); else process.stdout.write(payload);
